src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -349,6 +352,291 @@ private Boolean decorationStep(final Structure.StructureBlock structureBlock)
+            building.addContainerPosition(pos);
+        }
+
+        //It will crash at blocks like water which is actually free, we don't have to decrease the stacks we have.
+        if (isBlockFree(block, block.getMetaFromState(blockState)))
+        {
+            return true;
+        }
+
+        @Nullable final ItemStack stack = BlockUtils.getItemStackFromBlockState(blockState);
+        if (stack == null || stack == ItemStack.EMPTY)
+        {
+            Log.getLogger().error("Block causes NPE: " + blockState.getBlock());
+            return false;
+        }
+
+        final List<ItemStack> itemList = new ArrayList<>();
+        itemList.add(stack);
+        if (job instanceof JobBuilder && ((JobBuilder) job).getStructure() != null
+              && ((JobBuilder) job).getStructure().getBlockInfo() != null && ((JobBuilder) job).getStructure().getBlockInfo().tileentityData != null)
+        {
+            itemList.addAll(getItemStacksOfTileEntity(((JobBuilder) job).getStructure().getBlockInfo().tileentityData));
+        }
+
+        for (final ItemStack tempStack : itemList)
+        {
+            if (tempStack != null && tempStack != ItemStack.EMPTY)
+            {
+                final int slot = worker.findFirstSlotInInventoryWith(tempStack.getItem(), tempStack.getItemDamage());
+                if (slot != -1)
+                {
+                    new InvWrapper(getInventory()).extractItem(slot, 1, false);
+                    reduceNeededResources(tempStack);
+                }
+            }
+        }
+
+        if (Configurations.builderBuildBlockDelay > 0 && block != Blocks.AIR)
+        {
+            setDelay(Configurations.builderBuildBlockDelay);
+        }
+
+        return true;
+    }
+
+    /**
+     * Defines blocks that can be built for free.
+     *
+     * @param block    The block to check if it is free.
+     * @param metadata The metadata of the block.
+     * @return true or false.
+     */
+    public static boolean isBlockFree(@Nullable final Block block, final int metadata)
+    {
+        return block == null
+                 || BlockUtils.isWater(block.getDefaultState())
+                 || block.equals(Blocks.LEAVES)
+                 || block.equals(Blocks.LEAVES2)
+                 || (block.equals(Blocks.DOUBLE_PLANT) && Utils.testFlag(metadata, 0x08))
+                 || block.equals(Blocks.GRASS)
+                 || block.equals(ModBlocks.blockSolidSubstitution);
+    }
+
+    /**
+     * Reduce the needed resources by the itemStack.
+     *
+     * @param stack stack to reduce.
+     */
+    public void reduceNeededResources(final ItemStack stack)
+    {
+        final AbstractBuilding workerBuilding = this.getOwnBuilding();
+        if (workerBuilding instanceof BuildingBuilder)
+        {
+            ((BuildingBuilder) workerBuilding).reduceNeededResource(stack, 1);
+        }
+    }
+
     private Boolean structureStep(final Structure.StructureBlock structureBlock)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this method 7, down to the maximum allowed 6. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') The Cyclomatic Complexity of this method "structureStep" is 20 which is greater than 15 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 26 to the 20 allowed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)